File: 240929/갈 수 있는 곳들/places-can-go.py
from collections import deque
import sys
sys.setrecursionlimit(10000 + 1)

n, k = map(int, input().split())
area = []
for i in range(n):
    area.append(list(map(int, input().split())))

visited = []
for i in range(n):
    visited.append([0] * n)

q = deque()
for i in range(k):
    x, y = map(int, input().split())
    q.append([x-1, y-1])
    visited[x-1][y-1] = 1

def in_range(x, y):
    if 0 <= x < n and 0 <= y < n:
        return True
    return False

def can_go(x, y):
    if in_range(x, y) and area[x][y] == 0 and not visited[x][y]:
        return True
    return False

def bfs():
    while q:
        tmp = q.popleft()
        x, y = tmp[0], tmp[1]

        dx, dy = [0, 1, 0, -1], [1, 0, -1, 0]

        for dr in range(4):
            nx, ny = x + dx[dr], y + dy[dr]

            if can_go(nx, ny):
                visited[nx][ny] = 1
                q.append([nx, ny])
                bfs()

bfs()

ans = 0
for row in visited:
    ans += sum(row)

print(ans)

File: 250223/갈 수 있는 곳들/places-can-go.cpp
#include <iostream>
#include <queue>
using namespace std;

int n, k;
int ans = 0;
int area[100][100];
bool visited[100][100];
queue<pair<int, int>> q;

int dx[] = { -1, 0,1,0 };
int dy[] = { 0,1,0,-1 };

bool InRange(int x, int y) {
    return 0 <= x && x < n && 0 <= y && y < n;
}

bool CanGo(int x, int y) {
    return InRange(x, y) && area[x][y] == 0 && !visited[x][y];
}

void BFS() {
    while (!q.empty()) {
        pair<int, int> v = q.front();
        int x = v.first, y = v.second;
        q.pop();

        for (int dr = 0; dr < 4; dr++) {
            int nx = x + dx[dr];
            int ny = y + dy[dr];

            if (CanGo(nx, ny)) {
                ans++;
                q.push({ nx, ny });
                visited[nx][ny] = true;
            }
        }
    }
}

int main() {
    cin >> n >> k;
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            cin >> area[i][j];

    for (int i = 0; i < k; i++) {
        int x, y;
        cin >> x >> y;

        if (!visited[x - 1][y - 1]) {
            ans++;
            q.push({ x - 1, y - 1 });
            visited[x - 1][y - 1] = true;
            BFS();
        }
    }

    cout << ans;
    return 0;
}